common/app/navigation/helpers/UrlHelpers.scala

```diff
@@ -72,7 +80,8 @@ object UrlHelpers {
 
     // INTCMP is passed as a separate param because people look at it in Google Analytics
     // It's set to the most specific thing (componentId) to maximise its usefulness
-    destination.url ? ("INTCMP" -> componentId) & ("acquisitionData" -> acquisitionData.toString)
+    val url = destination.url ? ("INTCMP" -> componentId) & ("acquisitionData" -> acquisitionData.toString)
+    url.toString(uriEncoder)
```

Comment:
if this is the whole URL, this would surely encode the structural characters of the URL too? eg hash and ? 
Given this is a query string encoder, it should encode only the `query` part in 
```
URI         = scheme ":" hier-part [ "?" query ] [ "#" fragment ]
```
i.e the bit between the `?` and the `#` or end of string.

Since we have already combined the query part with the rest of the URL, it is too late.  We must do the encoding at the point we build the URI (and you'd hope that would be the whole purpose of the `com.netaporter.uri.dsl` library rather than it leaving it up to us

Reply:
this quote from T Berners Lee says it more accurately than me from 2.4.2 of https://www.ietf.org/rfc/rfc2396.txt

> A URI is always in an "escaped" form, since escaping or unescaping a
   completed URI might change its semantics.  Normally, the only time
   escape encodings can safely be made is when the URI is being created
   from its component parts; each component may have its own set of
   characters that are reserved, so only the mechanism responsible for
   generating or interpreting that component can determine whether or
   not escaping a character will change its semantics. Likewise, a URI
   must be separated into its components before the escaped characters
   within those components can be safely decoded.

Reply:
it looks like there is a final stage of escaping after all the assembly level escaping, all characters except the reserved and unreserved characters must be escaped, regardless of where they occur.
```
      reserved    = ";" | "/" | "?" | ":" | "@" | "&" | "=" | "+" |
                    "$" | ","

      unreserved  = alphanum | mark

      mark        = "-" | "_" | "." | "!" | "~" | "*" | "'" | "(" | ")"
```
This would include the quote, all non printable characters, space and <, > etc.

Looking at the library we are using, there is a more watertight variable we can use: https://github.com/lemonlabsuk/scala-uri/blob/0a7c66c864cf68298fa25131db24369d76a3c5de/shared/src/main/scala/io/lemonlabs/uri/encoding/PercentEncoder.scala#L51
although looking at the line above they do say "Probably more than you need to percent encode. Wherever possible try to use a tighter Set of characters to encode depending on your use case"

Why they didn't just implement a proper builder that does the encoding as you specifically add the components structurally is totally beyond me!!  Then they could read and implement the spec properly, and we wouldn't have to spend our time reading it.

Reply:
yeah I find that confusing too - from Tom's example in the PR description, it clearly does leave the `&` and the `?` intact, but reading https://github.com/lemonlabsuk/scala-uri/blob/master/shared/src/main/scala/io/lemonlabs/uri/encoding/PercentEncoder.scala#L34 you wouldn't think that was the case

Reply:
actually Joe it looks like the encoder you pass in is passed down the tree for each component of the URL to use as it sees fit.  When it gets to the query string encoder, it calls encode on the key and value specifically separately, which is actually correct:
https://github.com/lemonlabsuk/scala-uri/blob/0a7c66c864cf68298fa25131db24369d76a3c5de/shared/src/main/scala/io/lemonlabs/uri/config/UriConfig.scala#L26
The problem is the later stage where it concatenates the parts together: https://github.com/lemonlabsuk/scala-uri/blob/0a7c66c864cf68298fa25131db24369d76a3c5de/shared/src/main/scala/io/lemonlabs/uri/Uri.scala#L624 and it needs to encode ALL non allowed characters either at or before this point, and doesn't.

Reply:
My understanding was that currently `UriConfig.default` is being used, and so my change simply modifies this to encode extra chars on the querystring only.

I notice the `Uri.scala` you link to above is very different to the one being used by frontend, so we're clearly using an old version. The one frontend is using has the following signature for toString:
`def toString(implicit c: UriConfig = UriConfig.default): String`
Meaning it was already using `UriConfig.default`

Reply:
To be precise, it was previously implicitly using:
`implicit def uriToString(uri: Uri)(implicit c: UriConfig = UriConfig.default): String = uri.toString(c)`
Which [still uses UriConfig.default](https://github.com/lemonlabsuk/scala-uri/blob/2dc30fbea4076e50f3d4129f062f03e81ebf6754/shared/src/main/scala/io/lemonlabs/uri/dsl/package.scala#L16)

Reply:
I'll remove `,` and `:` as you're right, these are not required, we just need to encode `"` in the querystring